README.md

rdfp <img src="man/figures/rdfp.png" width="120px" align="right" />
===================================================================

[![Build Status](https://travis-ci.org/StevenMMortimer/rdfp.svg?branch=master)](https://travis-ci.org/StevenMMortimer/rdfp) [![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/StevenMMortimer/rdfp?branch=master&svg=true)](https://ci.appveyor.com/project/StevenMMortimer/rdfp) [![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/rdfp)](http://cran.r-project.org/package=rdfp) [![Coverage Status](https://codecov.io/gh/StevenMMortimer/rdfp/branch/master/graph/badge.svg)](https://codecov.io/gh/StevenMMortimer/rdfp?branch=master)

**Compiled using DFP API version: `v201802`**

**rdfp** allows you to use the DoubleClick for Publishers API from R. Manage inventory, create orders, pull reports, and more!

Table of Contents
-----------------

-   [Installation](#installation)
-   [Vignettes](#vignettes)
-   [Usage](#usage)
    -   [Functions](#functions)
    -   [Set API Version](#set-api-version)
    -   [Authenticate](#authenticate)
    -   [Get Current User Info](#get-current-user-info)
    -   [Pull a LineItem](#pull-a-lineitem)
    -   [Run a Report](#run-a-report)
-   [Credits](#credits)
-   [More Information](#more-information)

Installation
------------

``` r
install.packages("devtools")

# or to get the latest version available on GitHub
devtools::install_github("StevenMMortimer/rdfp")
```

If you encounter a clear bug, please file a minimal reproducible example on [GitHub](https://github.com/StevenMMortimer/rdfp/issues).

Vignettes
---------

The README below outlines the package functionality, but review the vignettes for more detailed examples on usage.

-   [Administrative Tasks](https://StevenMMortimer.github.io/rdfp/articles/administrative-tasks.html)
-   [Ad Trafficking Setup](https://StevenMMortimer.github.io/rdfp/articles/ad-trafficking-setup.html)
-   [Manipulating Orders and LineItems](https://StevenMMortimer.github.io/rdfp/articles/manipulating-orders-and-lineitems.html)
-   [Checking Availability](https://StevenMMortimer.github.io/rdfp/articles/checking-availability.html)
-   [Pulling Reports](https://StevenMMortimer.github.io/rdfp/articles/pulling-reports.html)

Usage
-----

### Functions

All functions start with `dfp_` so that you can easily identify DFP-specific operations and use tab completion in RStudio. Most **rdfp** functions will return a `data.frame` or `list` parsed from the XML returned in the SOAP response.

### Load Package and Set API Version

Google releases 4 versions of the DFP API each year and deprecates versions older than one year, so there is a lot that's regularly changing in this package. **rdfp** is periodically compiled against a recent version of the API. If you would like to use an older or newer API version that what is the package default, then just specify it as an option.

``` r
library(rdfp)
# see the package default version
getOption("rdfp.version")
#> [1] "v201802"
# this will force all calls to be against the version "v201711"
options(rdfp.version = "v201711")
```

### Authenticate

To authenticate you will first need to specify the `network_code` of the DFP instance you'd like to connect to. This is the only required option that the user must specify when using the **rdfp** package. After setting the `network_code` all you need to do is run `dfp_auth()`. If you already have a cached `.httr-oauth` file in the current working directory, then the token will be loaded and refreshed if necessary. Otherwise, your browswer will pop open and you will interactively authenticate.

The package has other options like a client\_id and client\_secret where you can connect using your own API client instead of the package default. Using your own client requires setting one up in the \[Google Developers Console\] (<https://console.developers.google.com>).

``` r
options(rdfp.network_code = "12345678")
options(rdfp.application_name = "MyApp")
options(rdfp.client_id = "012345678901-99thisisatest99.apps.googleusercontent.com")
options(rdfp.client_secret = "Th1s1sMyC1ientS3cr3t")

# this function will use the options set above and 
# cache an OAuth token in the working directory
# the token will be refreshed when necessary
dfp_auth()
```

### Get Current User Info

``` r
# Check current user or network
user_info <- dfp_getCurrentUser()
user_info[,c('id', 'isActive')]
#> # A tibble: 1 x 2
#>           id isActive
#>        <dbl> <chr>   
#> 1 185549536. true
network_info <- dfp_getCurrentNetwork()
network_info[,c('id', 'networkCode')]
#> # A tibble: 1 x 2
#>        id networkCode
#>     <dbl>       <dbl>
#> 1 109096.    1019096.
```

#### Pull a LineItem

The function `dfp_getLineItemsByStatement()` function from the [LineItemService](https://developers.google.com/doubleclick-publishers/docs/reference/v201802/LineItemService) allows you to retrieve Line Items by Publishers Query Language (PQL) statement. The statement is constructed as a list of lists that are nested to emulate the hierarchy of the XML that needs to be created in the request.

``` r
# Retrieve up to 3 Line Items that have a status of "DELIVERING"
request_data <- list('filterStatement'=list('query'="WHERE status='DELIVERING' LIMIT 3"))
resultset <- dfp_getLineItemsByStatement(request_data, as_df=TRUE) 
resultset[,c('orderId', 'id', 'priority', 'deliveryRateType')]
#> # A tibble: 3 x 4
#>       orderId          id priority deliveryRateType
#>         <dbl>       <dbl>    <dbl> <chr>           
#> 1 1254227776. 1928427256.      12. EVENLY          
#> 2 1393728736. 2038280896.      12. EVENLY          
#> 3 1265251936. 2060032216.      16. EVENLY
```

### Run a Report

Below is an example of how to make a simple report request.

``` r
# In order to run a report you must specify how the report should be structured 
# by specifying a reportQuery inside a reportJob. All of the dimensions, columns, 
# date range options, etc. are documented at:
# https://developers.google.com/doubleclick-publishers/docs/reference/v201802/ReportService.ReportQuery
request_data <- list(reportJob=list(reportQuery=list(dimensions='MONTH_AND_YEAR', 
                                                     dimensions='AD_UNIT_ID',
                                                     adUnitView='FLAT',
                                                     columns = 'AD_SERVER_IMPRESSIONS', 
                                                     columns = 'AD_SERVER_CLICKS',
                                                     dateRangeType='LAST_WEEK'
                                                     )))

# a convenience function has been provided to you to manage the report process workflow
# if you would like more control, see the example below which moves through each step in the process
report_data <- dfp_full_report_wrapper(request_data)
report_data[,c('Dimension.MONTH_AND_YEAR', 'Dimension.AD_UNIT_ID', 'Column.AD_SERVER_CLICKS')]
#> # A tibble: 25 x 3
#>    Dimension.MONTH_AND_YEAR Dimension.AD_UNIT_ID Column.AD_SERVER_CLICKS
#>    <chr>                                   <dbl>                   <dbl>
#>  1 2018-03                            133765096.                   1778.
#>  2 2018-03                            133765216.                    422.
#>  3 2018-03                            133765336.                    642.
#>  4 2018-03                            133765456.                   2176.
#>  5 2018-03                            142461136.                     91.
#>  6 2018-03                            133765576.                  17581.
#>  7 2018-03                            133765696.                    945.
#>  8 2018-03                            133765936.                  18171.
#>  9 2018-03                            142461016.                    312.
#> 10 2018-03                            143543776.                   1496.
#> # ... with 15 more rows
```

Credits
-------

This application uses other open source software components. The authentication components are mostly verbatim copies of the routines established in the **googlesheets** package (<https://github.com/jennybc/googlesheets>). We acknowledge and are grateful to these developers for their contributions to open source.

More Information
----------------

Google provides support for client libraries [here](https://developers.google.com/doubleclick-publishers/docs/clients), but unfortunately, R is not a supported language. Google's client libraries directly reference the production WSDLs to interact with the API, but this package makes SOAP requests best formatted to match the WSDL standards. This articulation is not perfect and continued progress will be made to bring functionality up to par with the client libraries.

Most all operations supported by the DFP API are available via this package. It is strongly recommended that you use the [DFP API Reference](https://developers.google.com/doubleclick-publishers/docs/rel_notes) when using this package. Details on formatting, attributes, and methods are all better explained by Google's documentation.

More information is also available on the `pkgdown` site at <https://StevenMMortimer.github.io/rdfp/>.

[Top](#rdfp-)

------------------------------------------------------------------------

Please note that this project is released with a [Contributor Code of Conduct](CONDUCT.md). By participating in this project you agree to abide by its terms.


NEWS.md
## rdfp 0.1.2

### Features

  * `dfp_report_url_to_dataframe()` now returns a `tbl_df` object for better 
  printing and casted data types
  * Simplify `dfp_select()` so that it automatically uses `dfp_select_parse()` 
  to format the result into a `tbl_df`
  * Add `dfp_date_as_list()` that will convert a date or datetime representation 
  in R to the required list format for submission to DFP
  * Add examples to functions so that the inline documentation 
  gives a better idea of how to make it work in R
  * Add `addtl_scopes` arugment to `dfp_auth()` so that users can create token that 
  can be used with other R packages to connect to other Google services

---

## rdfp 0.1.1 [release](https://github.com/StevenMMortimer/rdfp/releases/tag/v0.1.1)

### Features

  * Upgraded to API version `v201802`
  * Created more efficient parsers `parse_soap_response()`. Warning: This will cause 
  some breaking changes because how certain results are returned may be different.

### Bug Fixes

  * Line item targeting does not come back as a matrix. It is preserved as a list 
  so it can be directly passed back into the ForecastService functions or LineItemService 
  to get detailed availability

---

## rdfp 0.1.0 [release](https://github.com/StevenMMortimer/rdfp/releases/tag/v0.1.0)

### Features

  * Simple Administrative Tools:
    * `dfp_getCurrentUser()`
    * `dfp_getCurrentNetwork()`
    * `dfp_createTeams()`
    * `dfp_createUsers()`
    * `dfp_createUserTeamAssociations()`
    * `dfp_createCompanies()`
    * `dfp_createContacts()`
  * Ad Trafficking Setup: 
    * `dfp_createLabels()`
    * `dfp_createCustomFields()`
    * `dfp_createCustomFieldOptions()`
    * `dfp_createCustomTargetingKeys()`
    * `dfp_createCustomTargetingValues()`
  * Manipulating Orders and LineItems: 
    * `dfp_createOrders()`
    * `dfp_getLineItemsByStatement()`
  * Forecasting and Reporting:
    * `dfp_full_report_wrapper()`
    * `dfp_getDeliveryForecast()`
    * `dfp_getAvailabilityForecast()`
    
### Bug Fixes

  * None.


R/dfp_auth.R
# rdfp authentication

# Adapted from googlesheets package https://github.com/jennybc/googlesheets
# Specifically https://github.com/jennybc/googlesheets/blob/72abc6b218c26eecb9f32d0519cca41c6174aab8/R/gs_auth.R

# Changed elements:
#  - the scope list points to the dfp authentication endpoint: https://www.googleapis.com/auth/dfp
#  - the function get_google_token() will attempt to refresh itself before making call instead 
#    of allowing httr to refresh if 401 error occurs
#  - renamed the function gs_auth to dfp_auth to be consistent with package endpoint

# Copyright (c) 2017 Jennifer Bryan, Joanna Zhao

# Licensed under MIT license.

# environment to store credentials
.state <- new.env(parent = emptyenv())

#' Check that token appears to be legitimate
#'
#' This unexported function exists to catch tokens that are technically valid,
#' i.e. `inherits(token, "Token2.0")` is TRUE, but that have dysfunctional
#' credentials.
#'
#' @keywords internal
is_legit_token <- function(x, verbose = FALSE) {
  
  if(!inherits(x, "Token2.0")) {
    if(verbose) message("Not a Token2.0 object.")
    return(FALSE)
  }
  
  if("invalid_client" %in% unlist(x$credentials)) {
    # check for validity so error is found before making requests
    # shouldn't happen if id and secret don't change
    if(verbose) {
      message("Authorization error. Please check client_id and client_secret.")
    }
    return(FALSE)
  }
  
  if("invalid_request" %in% unlist(x$credentials)) {
    # known example: if user clicks "Cancel" instead of "Accept" when OAuth2
    # flow kicks to browser
    if(verbose) message("Authorization error. No access token obtained.")
    return(FALSE)
  }
  
  TRUE
  
}

#' Authorize \code{rdfp}
#'
#' Authorize \code{rdfp} to access your Google user data. You will be
#' directed to a web browser, asked to sign in to your Google account, and to
#' grant \code{rdfp} access to user data for Double Click for Publishers. 
#' These user credentials are cached in a file named
#' \code{.httr-oauth} in the current working directory, from where they can be
#' automatically refreshed, as necessary.
#'
#' Most users, most of the time, do not need to call this function
#' explicitly -- it will be triggered by the first action that
#' requires authorization. Even when called, the default arguments will often
#' suffice. However, when necessary, this function allows the user to
#'
#' \itemize{
#'   \item store a token -- the token is invisibly returned and can be assigned
#'   to an object or written to an \code{.rds} file
#'   \item read the token from an \code{.rds} file or pre-existing object in the
#'   workspace
#'   \item provide your own app key and secret -- this requires setting up a new
#'   project in
#'   \href{https://console.developers.google.com}{Google Developers Console}
#'   \item prevent caching of credentials in \code{.httr-oauth}
#' }
#'
#' In a call to \code{dfp_auth}, the user can provide the token, app key and
#' secret explicitly and can dictate whether credentials will be cached in
#' \code{.httr_oauth}. They must be specified.
#'
#' To set options in a more persistent way, predefine one or more of
#' them with lines like this in a \code{.Rprofile} file:
#' \preformatted{
#' options(rdfp.network_code = "12345678",
#'         rdfp.application_name = "MyApp",
#'         rdfp.client_id = "012345678901-99thisisatest99.apps.googleusercontent.com",
#'         rdfp.client_secret = "Th1s1sMyC1ientS3cr3t",
#'         rdfp.httr_oauth_cache = FALSE)
#' }
#' See \code{\link[base]{Startup}} for possible locations for this file and the
#' implications thereof.
#'
#' More detail is available from
#' \href{https://developers.google.com/identity/protocols/OAuth2}{Using OAuth
#' 2.0 to Access Google APIs}.
#'
#' @param token an actual token object or the path to a valid token stored as an
#'   \code{.rds} file
#' @param new_user logical, defaults to \code{FALSE}. Set to \code{TRUE} if you
#'   want to wipe the slate clean and re-authenticate with the same or different
#'   Google account. This deletes the \code{.httr-oauth} file in current working
#'   directory.
#' @param addtl_scopes character, strings that indicate additional Google services 
#' the client should authorize. Use this when trying to generate a token that will 
#' work to authenticate to other packages using Google services, such as the googlesheets 
#' package or RGoogleAnalytics package.
#' @param key,secret the "Client ID" and "Client secret" for the application
#' @param cache logical indicating if \code{rdfp} should cache
#'   credentials in the default cache file \code{.httr-oauth}
#' @param verbose a logical indicating if messages should be printed
#' @return an OAuth token object, specifically a
#'   \code{\link[=Token-class]{Token2.0}}, invisibly
#'
#' @export
dfp_auth <- function(token = NULL,
                     new_user = FALSE,
                     addtl_scopes = c("https://spreadsheets.google.com/feeds",
                                      "https://www.googleapis.com/auth/drive",
                                      "https://www.googleapis.com/auth/spreadsheets",
                                      "https://www.googleapis.com/auth/presentations",
                                      "https://www.googleapis.com/auth/analytics",
                                      "https://www.googleapis.com/auth/yt-analytics.readonly",
                                      "https://www.googleapis.com/auth/gmail.readonly", 
                                      "https://www.googleapis.com/auth/gmail.compose", 
                                      "https://www.googleapis.com/auth/gmail.send"),
                     key = getOption("rdfp.client_id"),
                     secret = getOption("rdfp.client_secret"),
                     cache = getOption("rdfp.httr_oauth_cache"), 
                     verbose = TRUE) {
  
  if(new_user && file.exists(".httr-oauth")) {
    if(verbose) message("Removing old credentials ...")
    file.remove(".httr-oauth")
  }
  
  if(is.null(token)) {
    
    scope_list <- c("https://www.googleapis.com/auth/dfp")
    if(!missing(addtl_scopes)){
      addtl_scopes <- match.arg(addtl_scopes, several.ok = TRUE)
      scope_list <- c(scope_list, addtl_scopes)
    }
    
    rdfp_app <- httr::oauth_app("google", key = key, secret = secret)
    
    google_token <-
      httr::oauth2.0_token(httr::oauth_endpoints("google"), rdfp_app,
                           scope = scope_list, cache = cache)
    
    stopifnot(is_legit_token(google_token, verbose = TRUE))
    
    .state$token <- google_token
    
  } else {
    
    if(is_legit_token(token)) {
      google_token <- token
    } else {
      google_token <- try(suppressWarnings(readRDS(token)), silent = TRUE)
      if(inherits(google_token, "try-error")) {
        if(verbose) {
         message(sprintf("Cannot read token from alleged .rds file:\n%s",
                         token))
        }
        return(invisible(NULL))
      } else if(!is_legit_token(google_token, verbose = TRUE)) {
        if(verbose) {
         message(sprintf("File does not contain a proper token:\n%s", token))
        }
        return(invisible(NULL))
      }
    }
    .state$token <- google_token
    
  }
  
  invisible(.state$token)
  
}

#' Retrieve Google token from environment
#'
#' Get token if it's previously stored, else prompt user to get one.
#' 
#' @note This function is meant to be used internally. Only use when debugging.
#' @keywords internal
#' @export
get_google_token <- function() {
  
  if(!is.null(.state$token) && !.state$token$validate()){
    this_config <- httr::config(token = .state$token$refresh())
  } else if(is.null(.state$token) || !is_legit_token(.state$token)) {
    dfp_auth()
    this_config <- httr::config(token = .state$token)
  } else {
    this_config <- httr::config(token = .state$token)
  }
  return(this_config)
}

#' Check if authorization currently in force
#'
#' @return logical
#'
#' @note This function is meant to be used internally. Only use when debugging.
#' @keywords internal
#' @export
token_exists <- function(verbose = TRUE) {
  
  if(is.null(.state$token)) {
    if(verbose) {
      message("No authorization yet in this session!")
      
      if(file.exists(".httr-oauth")) {
        message(paste("NOTE: a .httr-oauth file exists in current working",
                      "directory.\n Run dfp_auth() to use the",
                      "credentials cached in .httr-oauth for this session."))
      } else {
        message(paste("No .httr-oauth file exists in current working directory.",
                      "Run dfp_auth() to provide credentials."))
      }
      
    }
    
    FALSE
    
  } else {
    
    TRUE
    
  }
  
}
